### PR TITLE
Add endpoint to delete current session

### PR DESCRIPTION
**Pull request by matthew-white**
This PR adds an endpoint to delete the current session, making it easy to log out without specifying the token of the current session in the URL. We will probably use this endpoint in Frontend, but I also imagine that some API users would find it convenient.

#### What has been done to verify that this works as intended?

New tests. I copied some of the tests of DELETE /v1/sessions/:token. I didn't copy all the tests, since the two endpoints call a shared function: I think the tests of DELETE /v1/sessions/:token provide good coverage of that shared behavior.

#### Why is this the best possible solution? Were any other approaches considered?

I created a shared function because I think there's enough logic shared between DELETE /v1/sessions/:token and the new DELETE /v1/sessions/current to warrant one. DELETE /v1/sessions/current ends up being quite short and looks very similar to GET /v1/sessions/restore.

I chose the path &hellip;/current to be similar to another endpoint we have, /v1/users/current.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

API users will be able to use the new endpoint. The DELETE /v1/sessions/:token endpoint was modified, but much of its code was copied as-is to the shared function. Existing tests continue to pass.

#### Before submitting this PR, please make sure you have:

- [x] run `make test-full` and confirmed all checks still pass OR confirm CircleCI build passes
- [x] verified that any code from external sources are properly credited in comments or that everything is internally sourced

### CODE REVIEW

**Comment by matthew-white** [comment]
This new endpoint should help with getodk/central#400.

File: lib/resources/sessions.js [hunk 1]
```diff
@@ -48,24 +48,30 @@ module.exports = (service, endpoint) => {
   service.get('/sessions/restore', endpoint((_, { auth }) =>
     auth.session.orElse(Problem.user.notFound())));
 
+  const logOut = (Sessions, auth, session) =>
+    auth.canOrReject('session.end', session.actor)
+      .then(() => Sessions.terminate(session))
+      .then(() => (_, response) => {
+        // revoke the cookie associated w the session, if the session was used to
+        // terminate itself.
+        // TODO: repetitive w above.
```
Comment (alxndrsn) [inline lib/resources/sessions.js:57]: what does this TODO imply?

File: test/integration/api/sessions.js [hunk 2]
```diff
@@ -269,6 +269,35 @@ describe('api: /sessions', () => {
...
+      const { body } = await service.delete('/v1/sessions/current')
+        .set('Authorization', `Bearer ${token}`)
+        .expect(200);
+      body.should.eql({ success: true });
+      await service.get('/v1/users/current')
+        .set('Authorization', `Bearer ${token}`)
+        .expect(401);
+    }));
+
+    it('should not allow app users to delete their own sessions', testService(async (service) => {
```
Comment (alxndrsn) [inline test/integration/api/sessions.js:291]: Why not?  How do app users log out?

File: docs/api.md [hunk 3]
```diff
@@ -403,6 +403,26 @@ Logging out is not strictly necessary for Web Users; all sessions expire 24 hour
...
+#### Logging out current session [DELETE /v1/sessions/current]
+
+This endpoint causes the current session to log itself out. Logging out is not strictly necessary for Web Users; all sessions expire 24 hours after they are created. But it can be a good idea, in case someone else manages to steal your token.
+
+Only the session that was used to authenticate the request is logged out. If the Actor associated with the session has other sessions as well, those are not logged out.
+
++ Request
+    + Headers
+
+            Authorization: Bearer lSpAIeksRu1CNZs7!qjAot2T17dPzkrw9B4iTtpj7OoIJBmXvnHM8z8Ka4QPEjR7
```
Comment (alxndrsn) [inline docs/api.md:415]: For web users, the token will also be sent as a cookie, and perhaps in more cases.  Would it be helpful to also allow supplying the token-to-delete via cookie?

File: lib/resources/sessions.js [hunk 4]
```diff
@@ -48,24 +48,30 @@ module.exports = (service, endpoint) => {
   service.get('/sessions/restore', endpoint((_, { auth }) =>
     auth.session.orElse(Problem.user.notFound())));
 
+  const logOut = (Sessions, auth, session) =>
+    auth.canOrReject('session.end', session.actor)
+      .then(() => Sessions.terminate(session))
+      .then(() => (_, response) => {
+        // revoke the cookie associated w the session, if the session was used to
+        // terminate itself.
```
Comment (alxndrsn) [inline lib/resources/sessions.js:56]: When do you foresee clients invalidating _other_ sessions?  Is there even any point in authorizing a request to invalidate a session?  If a client knows a session token, then isn't that enough authority to invalidate that session?

File: docs/api.md (hunk 3 above)
Comment (matthew-white) [reply docs/api.md:415]: The token could be supplied via a cookie. Frontend might do that. However, we don't encourage API users to use cookie auth.

File: test/integration/api/sessions.js (hunk 2 above)
Comment (matthew-white) [reply test/integration/api/sessions.js:291]: App users auth using a session token. When access is revoked for an app user, that session is deleted. Project managers can revoke access for app users, but app users cannot do so themselves. App users are granted very limited rights and are often shared by multiple people.

File: lib/resources/sessions.js (hunk 4 above)
Comment (matthew-white) [reply lib/resources/sessions.js:56]: DELETE /v1/sessions/{token} is used for multiple things. One of them is for logging yourself out, but another is revoking access for app users and public links. See the comment above about app users. Project managers can revoke app users' sessions, but someone with a token for an app user or public link shouldn't be able to revoke its access.

On the other hand, DELETE /v1/sessions/current will probably just be used for logout.

File: lib/resources/sessions.js (hunk 1 above)
Comment (matthew-white) [reply lib/resources/sessions.js:57]: Heh I'm actually not 100% sure, mostly I'm just copying it over. I'm guessing that what's repetitive is the cookie attributes, which are repeated in the session create endpoint and the session delete endpoint.

I'd be happy to try to resolve this comment by creating a function to return these two cookies with the correct attributes. Both endpoints would call that function.

File: docs/api.md [hunk 5]
```diff
@@ -403,6 +403,26 @@ Logging out is not strictly necessary for Web Users; all sessions expire 24 hour
 + Response 403 (application/json)
     + Attributes (Error 403)
 
+#### Logging out current session [DELETE /v1/sessions/current]
+
+This endpoint causes the current session to log itself out. Logging out is not strictly necessary for Web Users; all sessions expire 24 hours after they are created. But it can be a good idea, in case someone else manages to steal your token.
```
[the remark this answers is not in the document]
Comment (matthew-white) [reply docs/api.md:408]: Mostly I was just copying this from the docs above for DELETE /v1/sessions/{token}. I guess once you're done with your API session, you might as well delete your token if you're not going to use it again. Or yeah, on a shared computer. I'd be happy to remove or reword this.

File: docs/api.md [hunk 6]
```diff
@@ -403,6 +403,26 @@ Logging out is not strictly necessary for Web Users; all sessions expire 24 hour
...
+
+            Authorization: Bearer lSpAIeksRu1CNZs7!qjAot2T17dPzkrw9B4iTtpj7OoIJBmXvnHM8z8Ka4QPEjR7
+
++ Response 200 (application/json)
+    + Attributes (Success)
+
++ Response 403 (application/json)
+    + Attributes (Error 403)
+
++ Response 404 (application/json)
```
Comment (matthew-white) [inline docs/api.md:423]: Most endpoints don't document the possibility of a 404, so I'll remove this.

**Review by alxndrsn — APPROVED** [review]

**Pull request merged**